packages/wast-parser/src/fsm.js

```diff
@@ -0,0 +1,98 @@
+// @flow
+
+const STOP = Symbol("STOP");
+
+type State<T> = T | Symbol;
+type TransitionEdge<Q> = () => [Q, number] | false;
+type TransitionFunction<Q> = () => [Q, number];
+type TransitionList<Q> = { [Q]: Array<TransitionEdge<Q>> };
+
+function makeTransition<Q>(
+  regex: RegExp,
+  nextState: Q,
+  n: number = 1,
+  allowedSeparator: ?string
+): TransitionEdge<Q> {
+  return function() {
+    if (allowedSeparator) {
+      if (this.input[this.ptr] === allowedSeparator) {
```

Comment:
The `this` looks a bit mystic to me. Could you rather pass the instance as argument in the function?

Reply:
Sadly not as easy, because the consuming code provides transitions as arguments to construct an `FSM` instance, so theres a bootstrapping problem. I solved it by binding all transition functions in `combineTransitions`, but its not the prettiest agreed. Couldnt think of a better solution yet.

Reply:
Sorry, I probably misunderstood. Your solution is indeed possible and I used it  9afce4c8a976a5ab4f23be519e4ac8a9a7da5fdc